### DISCUSSION TITLE
Stringify parser tree back to wikitext?

### DISCUSSION THREAD

**Discussion by linonetwo** [TiddlyWiki/TiddlyWiki5#6509] [Ideas]
I've successfully get a parse tree using the technique in [Use parsers as library #6141](https://github.com/Jermolene/TiddlyWiki5/discussions/6141) , and wikiast is making my WYSIWYG prototype work.

Now I need to turn wikiast back to wikitext, is there an existing method to do this? Wikiast contains node like `{ type: element, tag: p }`, this makes serializing complex a bit than serializing mdast.

**Comment by Jermolene** [comment]
Hi @linonetwo no, we do not have a way to convert an AST back to wikitext, but it's something I'd certainly like to see in the core.

**Reply by linonetwo** [reply]
I have written a basic one in https://github.com/tiddly-gittly/slate-write/tree/master/src/transform 
([wikiast-util-to-wikitext](https://github.com/tiddly-gittly/slate-write/tree/master/src/transform/wikiast-util-to-wikitext))


https://user-images.githubusercontent.com/3746270/156891431-db9a6b6a-d7b2-408a-9fa2-553ad8bfbeee.mov



I think this might not be very useful for most of the TW users, and it is written in TS to simplify the debugger process, so I'll publish this to npm after it's stable, instead of PR to the tw core.

I learn the file structure from [inokawa/remark-slate-transformer](https://github.com/inokawa/remark-slate-transformer) (and [syntax-tree/mdast-util-to-markdown](https://github.com/syntax-tree/mdast-util-to-markdown) but it is too complicated, contains functions without comment that I don't know the purpose)

Its code might not be the best, but I don't know how to make it better now. The "Compilers: Principles, Techniques, and Tools" didn't write much about how to organize the code of AST to text program.
